### ISSUE TITLE
update nested object

### ISSUE THREAD

**Issue by venelin-mihaylov** [hasura/graphql-engine#1573]
First, thank you very much for your work on Hasura, its a great product.
There is a feature that I would be very happy to have, wanted to know what are your thoughts on it.
Currently there Is a nested insert feature. However having only an insert without update is not very useful.

Do you have plans to implement update for nested objects? 
IMO, the logic would have to be:
I. for lists (1:N)
 - for every relation record:
    1. If it is present in input, not present in the db, then insert the record and link it to owning record
    2. If it is present in db, not present in input, the delete it
    3. If it is present in db and is present in input, then update its columns

II. For objects (belongsTo)
 - if relation record has an ID, link it to master and update its values
 - if relation record has no ID, then create it
 - if relation record is not present in input, but is present in db, then unlink it (an option to delete it would be nice)

Thank you

**Comment by alaarihan** [comment]
@venelin-mihaylov I have some notes about your suggestion:

> I. for lists (1:N)
> 
> * for every relation record:
>   
>   1. If it is present in input, not present in the db, then insert the record and link it to owning record
>   2. If it is present in db, not present in input, the delete it
in your delete case we will always have to include all of the related object to avoid deleting them! what if we have a million items, we need to include all of them to the nested update mutation to avoid losing them? I think we need another way for delete nested items
>   3. If it is present in db and is present in input, then update its columns
And we need a way to link existing items to the master without updating their data 
> 
> II. For objects (belongsTo)
> 
> * if relation record has an ID, link it to master and update its values
> * if relation record has no ID, then create it
> * if relation record is not present in input, but is present in db, then unlink it (an option to delete it would be nice)
Here is the same note too we can't delete all items when we don't include them in the nested mutation, we need to think in a different way.

**Comment by RodolfoSilva** [comment]
Any update about this issue?

**Comment by anasvn** [comment]
Waiting for this feature to reduce multiple hasura calls.

**Comment by nolandg** [comment]
@alaarihan The way Prisma solves this is with a `connect` and a `disconnect` field so you don't need to include a million ids in your mutation request. eg:

````gql
updatePost(_set: {
  tags: {
    connect: [
      {id: 1},
      {id: 2},
    ],
    disconnect: [
      {id: 3},
      {id: 4}
    ],
    insert: [
      {name: "cats"}
    ],
  }
})
````

Anything not mentioned in these fields is not touched. This works really well for me on Prisma. I'm trying to see how to convert my framework over to Hasura...will require some extra queries me thinks.

**Comment by alaarihan** [comment]
Yes @nolandg , I'm familiar with Prisma system but if we make this change in Hasura will make a big breaking change to the system so I was thinking in a different way to make it without breaking changes if it possible.

**Comment by luke-j** [comment]
Curious if this is on the roadmap? This seems like a really important feature to have, in my opinion.

**Comment by bkstorm** [comment]
yeah, it's inconvenient when hasura doesn't support this feature.
For example: a product can have a thumbnail (optional and is stored in image table), id of the thumbnail is stored in the field `thumbnail_id` in `product` table. 
First, I create a product without the thumbnail. It's ok, very easy.
Then I want to add a thumbnail to the product.
I have to insert a new row into `image` table, get `id` back then update `product` table with that `id`.
And all the operation should be in the same transaction.
What should I do, using event trigger or create a new resolvers on my own graphql server.

**Comment by RodolfoSilva** [comment]
Any update about this issue?

**Comment by switz** [comment]
Also would be very interested in a feature like this 👍

**Comment by rille111** [comment]
Would be very interested. Right now I just delete all nested relations and then just loop through what changes are made and insert them all. Ugly but works.

**Comment by heliocaruccio** [comment]
Hello developers. I'm in a part of development where it required a little bit of code dirt to perform nested updates. My question is whether this functionality will be developed and more or less when it might be made available to us. Depending on the case, I push certain features further.

**Comment by marionschleifer** [comment]
@heliocaruccio we are in the process of working on a spec for this issue. It should be ready within a month. We'll be able to share new information then 🙂

**Comment by rille111** [comment]
> @heliocaruccio we are in the process of working on a spec for this issue. It should be ready within a month. We'll be able to share new information then 🙂

Awesome! All love to the Hasura team. You make development with this tech fun.

**Comment by jfarcode** [comment]
Is there any news?

**Comment by marionschleifer** [comment]
We have published an RFC for this. You can find it here: https://github.com/hasura/graphql-engine/pull/3852.

**Comment by AntonBramsen** [comment]
Looking very much forward to this getting implemented.

Until then, does an example showing how to currently update relations between to objects exist?

To give some context, this is what I would like to achieve:

```graphql
mutation {
  update_user(where: { id: { _eq: "user-1" }}, 
    _set: {
      teams: {
        connect: [{ id: "team-1" }],
      }
    } 
  )
} 
```

**Comment by langfors** [comment]
Begging for this feature!  I was sad to see this not make it into 1.3.  We are drowning in tens of thousands of lines of code to manage deeply nested object updates and this fact sadly reduces the value of GraphQL in general for us.  The child delete & re-insert pattern of nested objects recommended by hasura is not only error prone but has also increased our postgresql server load when saving large object trees as dozens of updates/reinserts are often needed in our use case.  This will be a huge feature for us to simplify our codebase and reduce potential bug surface.

**Comment by irfansidespin** [comment]
can anyone help me?, these marked variables not working, if I am passing hardcode values instead, that's working fine.
![image](https://user-images.githubusercontent.com/70980786/93144486-85cd2c80-f703-11ea-8830-ca340234fb95.png)

**Comment by riyazq** [comment]
Any update on this? I really need this feature.

**Comment by zewa666** [comment]
this would be super helpful. I've meanwhile built something partially generic for one of my apps which goes along these lines, if somebody is looking for an idea on how to approach it via js

```javascript
// my custom list-editor ui component flags elements with a soft-delete and otherwise added/modified state
// so I get back an array of adaptions to perform
const project = { data: [{ name: "a", id: 1, __state__: "modified" }, { name: "d", id: 2, __state__: "added" }, { name: "c", id: 3, __state__: "deleted" }], ... };

// creating resulting mutations through a helper
const {
  added: tasksAdded,
  modified: tasksModified,
  deleted: tasksDeleted
} = mapTypeToMutations(project, {
  type: "tasks",
  pkColumnName: "id",
  insertOperation: "insert_tasks",
  updateOperation: "update_tasks_by_pk",
  deleteOperation: "delete_tasks"
});

// shoot them in one go
const query = gql`mutation {
      update_project_by_pk(
        pk_columns: { id: ${project.id} },
        _set: {
          updated_at: "${new Date().toISOString()}",
          title: "${project.title}"
        }
      ) {
        id
      }
  ${tasksDeleted}
  ${tasksModified}
  ${tasksAdded}
}`;
```

**Comment by dhruvbhatia7** [comment]
Any updates? Stuck on this

**Comment by yasinarik** [comment]
Yes. I have just realized that I can't update as an object whereas I can insert.

I use the version `v1.3.3-cloud.3`.

### This will be super helpful for such a situation (A basic relationship table):
```dart
PK id;
FK post_id;
FK tag_id;
```

Let's assume that I want to edit a post. I will remove some tags (objects) and add some others. Some of them will stay the same. 

I think I can do this only by using multiple Hasura mutations. Are any other ways possible at the moment?

Still waiting for a solution. Thank you so much for the hard-working Hasura team! :)

**Comment by mikejcooper** [comment]
I've been struggling with nested updates/upserts. The work around I'm using is: 

    Parent
	    PK id;
	    FK child_id;
	   	value;
    Child
	    PK id;
	    value;

> Mutation
```
  mutation UpdateNestedChild {
    insert_parent_one(
      object: {
        value: "apples updated"
        child: {
          data: { value: "pears updated", id: 17 }
          on_conflict: { constraint: child_pkey, update_columns: [value] }
        }
        id: 16
      }
      on_conflict: { constraint: parent_pkey, update_columns: [value, child_id] }
    ) {
      id
      value
      child {
        id
        value
      }
      child_id
    }
  }
```

> Result

    {
      "data": {
        "insert_parent_one": {
          "id": 16,
          "value": "apples updated",
          "child": {
            "id": 17,
            "value": "pears updated"
          },
          "child_id": 17
        }
      }
    }

**Comment by TaylorFacen** [comment]
Following this thread because I would love to be able to upsert nested objects in one call.

**Comment by mitar** [comment]
There is a button in the right sidebar where you can follow an issue without having to make a comment, "subscribe".

**Comment by healiseu** [comment]
That's just my two pennies' worth; generally speaking, isn't that issue related to how you can pass results from one query to another using variables ? Isn't that also a limitation in GraphQL (and not elegant at all in SQL) ? Eventually you end up using your favorite programming language to do that...

**Comment by Eitan1112** [comment]
Any status update?
Even when using the hack of updating with insert and using the on_conflict, it doesn't delete nested objects that are no longer passed in the variables, which makes the app logic for updating nested object so very complex

**Comment by jdgamble555** [comment]
Here is how neo4j does it: https://neo4j.com/docs/graphql-manual/current/mutations/update/

This is one of those very important updates that seems to be ignored for to years?  There is a current pull request since August... I am wondering what happened with that?

https://github.com/hasura/graphql-engine/pull/3852

J

**Comment by WGTW** [comment]
This feature request is **highly** anticipated; what are the current plans? It's been nearly three (!) years.

Thanks!

**Comment by sassela** [comment]
hey folks, thanks for your patience! We can see it's highly anticipated so we plan to revisit [this RFC](https://github.com/hasura/graphql-engine/pull/3852) and continue speccing this feature out soon. We'll keep this issue updated with progress or questions as we have them.

**Comment by DoisKoh** [comment]
This is probably not helpful but I'm just chiming in to also say that I really, really, really want this feature. Haha.

Thanks for the great product by the way...

**Comment by 13banda** [comment]
3 year so on. how many more year need to release this features?
look like hasura focus shifted to cloud.

**Comment by Bessonov** [comment]
Calm down, guys. Hasura is open source for relevant parts. Who holds you down from working on RFC and implementation?

**Comment by rccc** [comment]
> Calm down, guys. Hasura is open source for relevant parts. Who holds you down from working on RFC and implementation?

@Bessonov You are right, hasura is open source, but once can says that they received funding : https://hasura.io/blog/seriesc-100m-graphql-for-everyone/...

**Comment by heliocaruccio** [comment]
Yes, calm is great... but I think there was a lot of attention diversion to implement on other banks like mySQL leaving the base image huge and blurring a basic feature that is implemented in many frameworks, for example active record on rails has been doing this for a long time time. About being opensource.. yeah but it's in Haskel doesn't make it much easier... but still hearts in this project.

**Comment by FakhruddinAbdi** [comment]
Unfortunately, many projects like (Hasura, Prisma, Firebase), start by solving basic problems But later, get stuck on handling and solving real problems.
When you check the project for the first time and play with it, it works great and it you choose it for your next project
But when you invest in, you will block on the such problems which don't receive any updates for years

**Comment by rccc** [comment]
What about a crowdfunding action to paid some developpers so that this wanted  features can be wrote finally ? ;)

**Comment by rccc** [comment]
Founder linkedin is here : https://www.linkedin.com/feed/update/urn:li:activity:6981287074997960704/

I think tho only solution is to write directly to him.

**Comment by beepsoft** [comment]
I asked about it at the Hasura Community Meeting yesterday, and they promised to come and give a response regarding this issue. 🤞

**Comment by bigzhu** [comment]
It's been three years since I waited for this feature!

**Comment by rccc** [comment]
Just for info, Graphjin do this natively, without waiting for years ...

https://graphjin.com/docs/more/guide-to-graphql/#nested-update

**Comment by beepsoft** [comment]
> I asked about it at the Hasura Community Meeting yesterday, and they promised to come and give a response regarding this issue. 🤞

Well, no response this time either ...

**Comment by FakhruddinAbdi** [comment]
Let's try to mention the members if they can update this thread
@rikinsk 
@rakeshkky 
@tirumaraiselvan

**Comment by rccc** [comment]
@marionschleifer  

Today i put a comment  on Rajoshi Ghosh linkedin account telling we are waiting this for years.

**Comment by dameleney** [comment]
Support for Nested Mutations is on our roadmap. We do not have a timeline at present. Meanwhile please continue to follow this Github issue for future updates. A comprehensive rfc that covers all use cases and limitations of the feature will be posted on the community soon. We welcome more detailed feedback from you and the community once we provide those details. 

Thank you for adding details about use cases, requirements, and possible limitations. We will review the following existing solutions that were referenced in this issue that I have listed below and provide our proposed solution in an rfc. In the meantime please add more details regarding use cases and elements of the feature you would like to see as part of our solutions.

https://graphjin.com/docs/more/guide-to-graphql/#nested-update
https://neo4j.com/docs/graphql-manual/current/mutations/update/ 

We believe this is a vital feature for the continued success of Hasura within the marketplace and it is a high priority for us. As I mentioned we will be posting a more comprehensive rfc soon. Here are some details of the current design proposals we are iterating on. We welcome your comments here as well as in the forthcoming rfc.
_______

## The current use cases we are targeting are:

- ​​insert/upsert (create new object and relate)
- update (update object and ensure relation)
- delete (unrelate and delete object)
- link (relate with an existing object)
- unlink (unrelate an existing object)
- replace (unlink existing and link new)

## Nested updates example:

```
author (id, name, address_id)
article (id, title, author_id)
address (id, street_address, city).
update_authors(
    _set, 
    _inc,
    where,
    _relationships: {
        # array relationship
        articles: {
            _insert: { data, on_conflict}
            _update: {_set, where, _relationships}
            _delete: BoolExp
            _link: BoolExp
            _unlink: BoolExp
        }
        # object relationship
        address: {
            _insert: { data, on_conflict} 
            _update: {_set, _relationships}  # Note that this update has no "where"
            _delete: true
            _link: SingleBoolExp  # Special kind of "where" that should return 1 result else error
            _unlink: true
        }
    }
)

```

## Details:

Relationships between objects are inferred via reference keys (foreign-keys or manual).

In the above example, for the articles array relationship the reference key is article.author_id -> author.id and for the address object relationship the reference key is author.address_id -> address.id.

- Insert/Upsert: Create new nested oject and set the reference key to establish the relationship
- Update: Update the nested object and update the reference key if required
- Delete: Delete the nested object and set the reference key to null if required (if nullable)
- Link: Set the reference key to establish the relationship
- Unlink: Set the reference key to null to remove the relationship (if nullable)
- Replace: Can be achieved via _unlink/_delete + _link/_insert combination

**Comment by nathikazad** [comment]
Yes please, its a struggle without this. For every relation we need to make a mutation and then update

**Comment by beepsoft** [comment]
@dameleney Is there any progress with the RFC and implementation?

Funny you mentioned graphjin (as an inspiration): they just seem to implemented transactions and they may have taken inspiration from an earlier Hasura RFC https://github.com/dosco/graphjin/issues/411

By the way: it would be great to have transactions as well. :-)

**Comment by nathikazad** [comment]
Ya this is insane, @coco98 please take a look into this. Its been 4 years.

**Comment by PedroBern** [comment]
Currently, the most effective approach I have found to create a new relationship within an update mutation is to utilize multiple mutations in a single request.

```graphql
mutation createFooAndUpdateBarByPK(
  $insert_object: foo_insert_input!
  $update_pk: uuid!
  $update_set: bar_set_input!
) {
  insert_foo_one(object: $insert_object) {
    id
    # Avoid requesting foo fields here as they may be outdated
    # (due to potential side effects on foo after linking it to bar)
  }
  # The "_set" field will include a reference to foo, such as foo_pk: "..."
  update_bar_by_pk(pk_columns: { pk: $pk }, _set: $update_set) {
   ...bar
   foo {
    ...foo # Request updated foo fields
   }
  }
}
```

**Comment by manasag** [comment]
Hi @beepsoft, we have now made Logical Models available for community edition with v2.28 onwards.

**Comment by FakhruddinAbdi** [comment]
Can anybody explain please how Logical Model will address the update nested object problem?

**Comment by beepsoft** [comment]
> Hi @beepsoft, we have now made Logical Models available for community edition with v2.28 onwards.

Thanks, @manasag, great news! And I see that also [native queries](https://hasura.io/docs/latest/schema/postgres/logical-models/native-queries/) are now supported for CE as well.

**Comment by FakhruddinAbdi** [comment]
> Can anybody explain please how Logical Model will address the update nested object problem?

After some research i got the answer

The Logical Model feature combined with Native Queries may potentially address nested updates. we can define a SQL query that performs the desired update operation on nested objects and expose this as a GraphQL mutation. 

For example, if we have a one-to-many relation from authors to posts, we can write a SQL query to update an author and their related posts, and expose this through Hasura's Native Queries. 

But, we should note that as of now, only read-only queries are supported, but future updates might allow mutations.

**Comment by manasag** [comment]
Yes, and I apologise for confusion from Hasura side. Logical Models (launched and available in CE in v2.2.8) currently only support queries and not mutations. Nested update can be solved once we support mutations with Logical Models. We have this in our roadmap; thanks for being patient on this issue.

**Comment by cbsmith402** [comment]
> Currently, the most effective approach I have found to create a new object and set the relationship within an update mutation is to utilize multiple mutations in a single request.
> 
> ```graphql
> mutation createFooAndUpdateBarByPK(
>   $insert_object: foo_insert_input!
>   $update_pk: uuid!
>   $update_set: bar_set_input!
> ) {
>   insert_foo_one(object: $insert_object) {
>     id
>     # Avoid requesting foo fields here as they may be outdated
>     # (due to potential side effects on foo after linking it to bar)
>   }
>   # The "_set" field will include a reference to foo, such as foo_pk: "..."
>   update_bar_by_pk(pk_columns: { pk: $pk }, _set: $update_set) {
>    ...bar
>    foo {
>     ...foo # Request updated foo fields
>    }
>   }
> }
> ```

@PedroBern Can you elaborate on this solution a bit? Are you saying that somehow the result of `insert_foo_one` is available to the `update_bar_by_pk` mutation?

**Comment by PedroBern** [comment]
> @PedroBern Can you elaborate on this solution a bit? Are you saying that somehow the result of insert_foo_one is available to the update_bar_by_pk mutation?

@cbsmith402 yes of course.

The mutations are executed in sequence, so yes, the result of the first mutation is already present in the database when the second one is executed.

In my example:
- The first mutation inserts foo
- The second mutation sets the relationship between foo and bar

Note that you must know the primary key of foo __before__ inserting it, in order to link it to bar __on a single request__ (or better say, on a single db transaction), see the updated code below:

```graphql
mutation createFooAndUpdateBarByPK(
  $foo_pk: uuid!
  $bar_pk: uuid!
) {
  # create foo
  insert_foo_one(object: {pk: $foo_pk}) {
    pk
  }
  # link foo and bar + request bar with related foo
  update_bar_by_pk(pk_columns: { pk: $bar_pk }, _set: { foo_pk: $foo_pk }) {
   ...bar
   foo {
    ...foo
   }
  }
}
```

**Comment by cbsmith402** [comment]
> Note that you must know the primary key of foo **before** inserting it, in order to link it to bar **on a single request** (or better say, on a single db transaction)

Ah, yep, I came to the same conclusion (UUIDs for the win!). Just wanted to make sure I wasn't missing something.

**Comment by jdgamble555** [comment]
Is this still not possible in 2024?

J

**Comment by vpamrit** [comment]
+1

Looking forward to this feature, thank you all!

**Comment by ahamid** [comment]
+1
